File: 0338-counting-bits/0338-counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        dp=[0]*(n+1)
        dp[0]=0
        for i in range(n+1):
            if(i%2==0):
                print(i//2)
                dp[i]=dp[i//2]
            else:
                dp[i]=dp[i//2]+1
        return dp

File: 0338-counting-bits/0338-counting-bits.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        // https://www.youtube.com/watch?v=awxaRgUB4Kw
        //Observe if the number is odd-> dividing by 2(>>1) has 1 less bit
        // if number if even then dividing by 2 doesnt change the number of set bits
        // odd has 1 in its LSB and even has 0
        
        vector<int> dp(n+1,0);
        
        dp[0]=0;
        
        for(int i=1;i<=n;i++){
            if(i%2==0){
                //even
                dp[i]=dp[i/2];
            }else{
                dp[i]=dp[i/2]+1;
            }
        }
        return dp;
    }
};